---
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-"
)
```

# brainSTORM

<!-- badges: start -->
`r badger::badge_devel("SchwartzLab/brainSTORM", "blue")`
<!-- badges: end -->

```{r}
# setwd("/home/labs/schwartzlab/miguelg/github_repos/brainSTORM/")
# rmTmpDir()
# unlink("bam_Pa", recursive = TRUE)
load_all()

# Genome and Gene Annotation
fastaGenome <- "/home/labs/schwartzlab/joeg/genomes/Pyrodictium_abyssi/prokka/PROKKA_01222021.fsa"
bedAnnotation <- "/home/labs/schwartzlab/joeg/genomes/Pyrodictium_abyssi/prokka/PROKKA_01222021_BED6.bed"
rRNA <- c("PYCC_00363", "PYCC_00364")
# fastaGenome <- "/home/labs/schwartzlab/miguelg/BIGDATA/genome_references/p_occultum/pAbys_txome.fa"
# bedAnnotation <- "/home/labs/schwartzlab/miguelg/BIGDATA/gene_annotations/p_occultum/pAbys_txome.bed"


# STAR Genomes generation
STARGenome <- mkSTARgenome(fastaGenome, bedAnnotation)
bisGPath <- bisGenome(fastaGenome)
STARGenome_bis <- mkSTARgenome(bisGPath, bedAnnotation)

# META
vList <- list(organism = c("PyroAbyss", "TherAcid", "Yeast", "Human"), 
              RTase = c("SSIII", "SSIV", "TGIRT"),
              libTreat = c("Ac4C", "CMC", "DeacetylatedAc4C", "MocklowdNTPs",
                           "Mock", "Dimroth", "m5C"),
              bioTreat = c("80deg", "95deg", "100deg", "AcidpH1", "AcidpH2",
                           "AcidpH3"))

r1Files <- file.path("fastq", grep(list.files("fastq"), pattern = "R1", value = TRUE))

META <- storm_META(fileNames = r1Files,
                      varsList = vList,
                      setVars = "bioTreat",
                      idVars = c("organism", "RTase", "libTreat", "bioTreat"),
                      groupVars = c("libTreat", "RTase"))

alignSTAR(read1Files = META[libTreat != "m5C", FASTQ], 
          nCores = 8, 
          zipped = TRUE,
          STARgenomeDir = STARGenome, 
          alignEndsType = "Local", 
          outSAMtype = "BAM Unsorted", 
          outDir = "bam_Pa")

alignSTAR(read1Files = META[libTreat == "m5C", FASTQ], 
          nCores = 8, 
          zipped = TRUE,
          STARgenomeDir = STARGenome_bis, 
          alignEndsType = "Local", 
          outSAMtype = "BAM Unsorted", 
          outDir = "bam_Pa")



```


