---
title: "Using_brainSTORM_at_WEXAC"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using_brainSTORM_at_WEXAC}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


```{r, Global Options, echo = F}
knitr::opts_chunk$set(warning = F, message = F, fig.pos = "center", fig.width = 10, fig.height = 8)
```

## SETUP

**Packages**

```{r}
library(brainSTORM)
library(txtools)
library(magrittr)
library(ggplot2)
library(parallel)
library(gridExtra)
```

These are the file paths needed to start the run:

```{r}
# Genome and Gene Annotation
fastaGenome <- "/home/labs/schwartzlab/miguelg/BIGDATA/genome_references/ribosomal/Sc_ribosomal_seqs.fa"
bedAnnotation <- "/home/labs/schwartzlab/miguelg/BIGDATA/gene_annotations/ribosomal/Sc_ribosomal_seqs.bed"
r1Files <- grep(list.files("/home/labs/schwartzlab/joeg/data/lib524/Saccharomyces_cerevisiae_SK1",
                           full.names = TRUE), pattern = "R1", value = TRUE)
OUTDIR <- "/home/labs/schwartzlab/miguelg/WORKSPACE_wexac/storm_seq/lib524/s_cerevisiae"
EXP_NAME <- "S.cerevisiae_SK1-lib524"
NCORES <- 1
```

## Design matrix from R1 Fasta files.

```{r}
vList <- list(organism = c("PyroAbyss", "TherAcid", "Yeast", "Human"), 
              RTase = c("SSIII", "SSIV", "TGIRT", "RTHIV"),
              libTreat = c("Ac4C", "CMC", "DeacetylatedAc4C", "MocklowdNTPs",
                           "Mock", "Dimroth", "m5C", "AlkBmix", "RBSseqHeatMg", "NaBH4HydBiotin"),
              bioTreat = c("80deg", "95deg", "100deg", "AcidpH1", "AcidpH2", "AcidpH3"))

META <- storm_META(fileNames = r1Files,
                   varsList = vList,
                   setVars = "organism",
                   idVars = c("organism", "RTase", "libTreat"),
                   groupVars = c("libTreat", "RTase"), 
                   outDir = OUTDIR)
```

## STAR genome and mapping reads

* Create transcriptome fasta and BED references from genomic versions
    + This is done so that bisulphite conversion is done accordingly to strand orientation.
* STAR Genomes generation

```{r, eval = FALSE}
# Make transcriptome, make new gene annotation for transcriptome
fastaTxOme <- mkTranscriptome(fastaGenome, bedAnnotation, nCores = NCORES)
bedTxOme <- mkBedFromFastaTxOme(fastaTxOme)

# Creating bisulphite transcriptome
bisTxPath <- bisGenome(fastaTxOme)

# Create STAR genomes
STARGenome <- mkSTARgenome(fastaTxOme, bedTxOme)
STARGenome_bis <- mkSTARgenome(bisTxPath, bedTxOme)

#Loading transcriptome and annotation
GENOME <- txtools::tx_load_genome(fastaTxOme)
TXOME <- txtools::tx_load_bed(bedTxOme)
```


* Read mapping to reference transcriptome using STAR

```{r, STAR mapping reads, eval = FALSE}
if(!all(file.exists(META$BAM))){
    # STAR Alignment
    alignSTAR(read1Files = META[libTreat != "m5C" & libTreat != "RBSseqHeatMg", FASTQ], 
              nCores = NCORES, 
              zipped = TRUE,
              STARgenomeDir = STARGenome, 
              alignEndsType = "Local", 
              outSAMtype = "BAM Unsorted",
              alignIntronMax = 1, #set to 1 for bacteria/archaea, adjust accordingly for eukaryotes!
              outDir = OUTDIR)
    alignSTAR(read1Files = META[libTreat == "m5C" | libTreat == "RBSseqHeatMg", FASTQ], 
              nCores = NCORES, 
              zipped = TRUE,
              STARgenomeDir = STARGenome_bis, 
              alignEndsType = "Local", 
              outSAMtype = "BAM Unsorted",
              alignIntronMax = 1, #set to 1 for bacteria/archaea, adjust accordingly for eukaryotes!
              outDir = OUTDIR)
}
```

## Processing to count data tables using txtools

* Process all BAM into TxDT

```{r, bam2txDT, eval = FALSE}
if(!all(file.exists(META$RDS))){
    rdsFiles <- parallel::mclapply(mc.cores = NCORES, seq_along(META$FASTQ), function(i){
        bam2TxDT(BAMfile = META$BAM[i], 
                 geneAnnot = TXOME, 
                 genome = GENOME, 
                 dtType = "covNuc", 
                 outDir = OUTDIR, 
                 nCores = 1, 
                 remL = 1000, 
                 minR = 0)
    })
}
```

## Reports

* Library complexity report

```{r, eval = FALSE}
libComplexReport(META, maxExtrapolation = 2.01e6, steps = 1e5, verbose = T)
```
```{r}
ggLCE <- gg_lce(META, tab_name = file.path(OUTDIR, "libCompReport.txt"), speciesName = EXP_NAME)
plot(ggLCE)
```

* Nucleotide frequency report

```{r}
ggNucF <- fastq_nucFreq(META, NCORES, firstN =  1e5) %>% 
    gg_nucFreq(subtitle = EXP_NAME)
plot(ggNucF)
```

* Alignment efficiency report

```{r}
rReport <- reads_report(META, NCORES)
ggReadStats <- gg_readStats(rReport, EXP_NAME)
gridExtra::grid.arrange(ggReadStats[[1]], ggReadStats[[2]], ncol = 2)
```

## brainSTORM

* STORM object creation
* Metrics calculation 
* Assignment to RNAmods tables

> Here the object `yeast_STORM` is an in-built dataset to work as an example,
which was created using the above code.

```{r}
# STORM <- storm_STORM(META, GENOME, TXOME, nCores = 1) 
STORM <- yeast_STORM %>% 
    add_default_metrics_v2() %>% 
    storm_makeCalls()
# Add known RNA modification sites
rRNAmods_Sc <- readRDS("/home/labs/schwartzlab/miguelg/BIGDATA/RNAmod_Annot/Taoka/rib_mods_Sc.rds")
STORM <- addKnownRNAmods(STORM, rRNAmods_Sc)
```

## Plotting scores by RNA modification

```{r, fig.width= 12, fig.height=8, fig.align="center"}
storm_metricsBoxPlot_byNuc(STORM, Y_scores, "Yeast - PseudoU scores")
storm_metricsBoxPlot_byNuc(STORM, Nm_scores, "Yeast - 2Ometh scores")
storm_metricsBoxPlot_byNuc(STORM, ac4C_scores, "Yeast - ac4C scores")
storm_metricsBoxPlot_byNuc(STORM, m7G_scores, "Yeast - m7G scores")
storm_metricsBoxPlot_byNuc(STORM, m1A_scores, "Yeast - m1A scores")
storm_metricsBoxPlot_byNuc(STORM, m5C_scores, "Yeast - m5C scores")
storm_metricsBoxPlot_byNuc(STORM, m3U_scores, "Yeast - m3U scores")
```

# Session Info

```{r, Session Info, echo = FALSE}
sessionInfo()
```
